### PR TITLE
[move] introduce bytecode specification

### PR DESCRIPTION
**Pull request by vgao1996**
## Description
This introduces a procedural macro `bytecode_spec` that allows us to add inline specifications to the Move bytecode instructions defined in Rust:
```rust
#[bytecode_spec]
pub enum Bytecode {
    #[group = "stack_and_local"]
    #[description = "Pop and discard the value at the top of the stack. The value on the stack must be an copyable type."]
    #[semantics = "stack >> _"]
    #[paranoid_post = r#"
        ty_stack >> ty
        assert ty has drop
    "#]
    Pop,

    ..
}
```
The macro will generate a `pub fn spec() -> Vec<BTreeMap<String, String>>`, which can be used to extract the specification defined using the the attributes. 
```json
[
    {
        "description": "Pop and discard the value at the top of the stack. The value on the stack must be an copyable type.",
        "group": "stack_and_local",
        "name": "pop",
        "paranoid_post": "ty_stack >> ty\nassert ty has drop",
        "semantics": "stack >> _",
    },
]
```
The extracted specification can then be used to generate other human readable docs (e.g. pdf, html, LaTeX etc.) The generator(s) will come in future PRs.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [x] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Unit tests

## Key Areas to Review
- `move-binary-format/src/file_format.rs`: full specification for all current bytecode instructions
- `move-bytecode-spec`: procedural macro to allow the specification to be extracted

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vgao1996) [inline third_party/move/move-binary-format/src/file_format.rs:1359]: Note: it's been some time since I wrote the semantics here so I'd like to double check it. Will get to it a bit later but will definitely get it done before landing.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (ziaptos) [inline third_party/move/move-binary-format/src/file_format.rs:1353]: I'd prefer not calling this `paranoid`. Maybe `runtime_type_check_epilogue` and `runtime_type_check_prologue`. Calling it just `paranoid` (not even `paranoid_type_check`).

**Review by ziaptos — CHANGES_REQUESTED** [review]
Looks straight-forward and good. But I'm concern about the `paranoid` field - see inline comment.

Also, should we stop using `move-X-Y-Z` directory names, as they are all obviously inside the `move` directory?

File: third_party/move/move-bytecode-spec/Cargo.toml
Comment (georgemitenkov) [inline third_party/move/move-bytecode-spec/Cargo.toml:5]: Description is wrong :)

File: third_party/move/move-bytecode-spec/tests/tests.rs
Comment (georgemitenkov) [inline third_party/move/move-bytecode-spec/tests/tests.rs:128]: Why these are dead? These are tests?

File: third_party/move/move-binary-format/src/file_format.rs
Comment (georgemitenkov) [inline third_party/move/move-binary-format/src/file_format.rs:1368]: typo: returned? Let's proof-read the comments before landing

File: third_party/move/move-binary-format/src/file_format.rs
Comment (georgemitenkov) [inline third_party/move/move-binary-format/src/file_format.rs:2590]: Maybe a nicer name? :)

**Review by georgemitenkov — APPROVED** [review]
Please address Zi's comment before landing

**Comment by georgemitenkov** [comment]
For directories, good question. We also have everything called `aptos-...`. Would be good to make this consistent (not prefixed at all?)

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vgao1996) [reply third_party/move/move-binary-format/src/file_format.rs:2590]: My bad, used this as an ad-hoc entry point to print out the extracted spec. Will get it removed.

File: third_party/move/move-bytecode-spec/tests/tests.rs
Comment (vgao1996) [reply third_party/move/move-bytecode-spec/tests/tests.rs:128]: This is to suppress the `never-constructed enum variants` warning. The compiler cannot see that the variants are being processed by the procedural macro.

**Comment by vgao1996** [comment]
@georgemitenkov @ziaptos Yeah the discussion on directory naming is great. Although I feel like we should handle it separately for the sake of consistency. Need to first reach agreement on a convention, and then perform batch renaming.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2318]: Are there any restrictions on the shifted value being `u64` (document it eitherway)?

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2332]: These do not seem to reflect what is being described in the semantics description.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2354]: Again, incorrect reflection of the semantics description.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2375]:
```suggestion
        elem_ty = instantiate elem_ty 
```

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2367]: In some places, we use `[u64_value]`, but here we use `[n]`?

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2394]: From the semantics description, it seems like this should be a reference to a vector type, not reference to an element of vector?

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2436]: Should the reference to the vector be `&mut` instead?

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2498]:
```suggestion
        Abort the execution if any of the indices are out of bounds.
```

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2522]: Weird that these show up down here instead of with the other `LdUx` above. Would be nice if they were together.

**Review by vineethk — COMMENTED** [review]
I really like this effort, and the PR looks generally pretty good to me.

I have a bunch of comments, will approve after they are addressed.

File: third_party/move/move-bytecode-spec/src/lib.rs
Comment (vgao1996) [reply third_party/move/move-bytecode-spec/src/lib.rs:131]: Not sure if we should do that as unknown attributes could belong to other macros. Stripping the ones you know should put us in a safer position.

Comment (vineethk) [reply third_party/move/move-bytecode-spec/src/lib.rs:131]: Ah, fair point. I was just worried about potential typos in the attribute names. Perhaps eventually, we can make some subset of these required attributes, and check that each variant has those.

File: third_party/move/move-bytecode-spec/src/lib.rs
Comment (vgao1996) [reply third_party/move/move-bytecode-spec/src/lib.rs:158]: For now just personal preference. I find "ld_u8" a bit more readable than "LdU8", as the latter is very cramped, making it hard for my brain to tokenize it.

Totally open for discussion though.

Comment (vineethk) [reply third_party/move/move-bytecode-spec/src/lib.rs:158]: I don't have a strong opinion either way, but having it be `LdU8` means I can copy it from code and search verbatim in whatever semantics document we produce (eg, in the html page) for it.

Comment (vineethk) [reply third_party/move/move-bytecode-spec/src/lib.rs:158]: I see that you used the lower_snake_case convention in the gas profiler output as well. So keeping this consistent is better - in light of this, I would suggest keeping this as is.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vgao1996) [reply third_party/move/move-binary-format/src/file_format.rs:1654]: Checked by the bytecode verifier.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vgao1996) [reply third_party/move/move-binary-format/src/file_format.rs:2332]: Fixed. The original paranoid mode code used the wrong notations.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vgao1996) [reply third_party/move/move-binary-format/src/file_format.rs:2394]: Yes, I'm just using the notion `&elem_ty` to indicate that this is a reference type created by prefixing `elem_ty` with `&`. 

Might be a bit confusing if you think of it in terms rust, but within the context of paranoid mode we only have types and no values so I guess it should be fine. Later we can take another look at the pseudocode syntax and see if we can improve it.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vgao1996) [reply third_party/move/move-binary-format/src/file_format.rs:2522]: The reason why these instructions appear at the end is because they were added at a later time. 

Now since we don't derive the serialization of the instructions it may very well be fine to change the order, but that's better done in a separate PR, with adequate verification.

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vgao1996) [reply third_party/move/move-binary-format/src/file_format.rs:1587]: In my original doc I have a line stating that `current_frame.pc += 1` is implicit for all non-control-flow instructions.

For the control-flow instructions however, changes to PC must be explicitly specified due to them being non-trivial.

**Comment by vgao1996** [comment]
@vineethk @georgemitenkov @ziaptos I've addressed all your comments except for the derivation of static operands -- this one requires a bit more hacking with parsing so I'll probably come back to it later.

**Comment by vgao1996** [comment]
@vineethk For static operands, I think it would be ideal to have a sub attribute `#[operand_name]` attached to each operand. 
```
LdU32(#[operand_name = "value_to_load"] u32)
```
This will be translated to `[value_to_load: u32]`

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [inline third_party/move/move-binary-format/src/file_format.rs:2154]: In the semantics, we pop off rhs, then lhs. In the runtime_check_epilogue, we pop off left_ty and then right_ty. This happens for several bytecodes and should be fixed.

**Review by vineethk — APPROVED** [review]
I have one other comment added (about the ordering of lhs and rhs in semantics vs. paranoid mode). Other than that, LGTM.

**Review by ziaptos — COMMENTED** [review]
LGTM

**Review by ziaptos — APPROVED** [review]
LGTM

File: third_party/move/move-binary-format/src/file_format.rs
Comment (vineethk) [reply third_party/move/move-binary-format/src/file_format.rs:2154]: @vgao1996 Just wanted to check if this was fixed.

Comment (vgao1996) [reply third_party/move/move-binary-format/src/file_format.rs:2154]: Not yet, but will fix this + check similar instructions before landing.

**Review by vgao1996 — COMMENTED** [review]
@vineethk I've fixed the lhs/rhs confusion, which originated from the paranoid mode code itself, which I also fixed. Now trying to land the PR

**Pull request merged**
